wasm/src/api.rs

```diff
@@ -40,3 +53,82 @@ pub fn run_clvm(program: &[u8], args: &[u8]) -> Vec<u8> {
         Err(_eval_err) => format!("{:?}", _eval_err).into(),
     }
 }
+
+#[wasm_bindgen]
+/**
+ * Return serialized result of clvm program with cost
+ *
+ * cost will be available at the first 8 bytes of returned Vec<u8>(=Uint8Array).
+ * bytes of node will be available from 8th byte offset of returned Vec<u8>.
+ *
+ * @example
+ * const result = run_chia_program(...); // Uint8Array
+ * const cost = vec_u8_to_u64(result.subarray(0, 8)) // BigInt;
+ * const serialized_node = result.subarray(8); // Uint8Array
+ */
+pub fn run_chia_program(
+    program: &[u8],
+    args: &[u8],
+    max_cost: Cost, // Expecting `BigInt` to be passed from JavaScript world
+    flag: u32,
+) -> Vec<u8> {
+    let mut allocator = Allocator::new();
+    let program = node_from_bytes(&mut allocator, program).unwrap();
+    let args = node_from_bytes(&mut allocator, args).unwrap();
+    let dialect = ChiaDialect::new(flag);
+
+    let r = run_program(
+        &mut allocator,
+        &dialect,
+        program,
+        args,
+        max_cost,
+        None,
+    );
+    match r {
+        Ok(reduction) => {
+            let cost = reduction.0;
+            let node_bytes = node_to_bytes(
+                &Node::new(&allocator, reduction.1)
```

Comment:
this is vulnerable to a DoS attack. It's possible to (cheaply) produce a return value that does not use a lot of memory in its internal clvm_rs representation, but that balloons into using a *lot* of memory when serialized. As long as only trusted programs are run, this is probably fine. There should probably be some kind of warning label though.

Reply:
As for the DoS attack you mentioned. could you correct me if my understanding is wrong?
My understanding:
- `SExp` instance on clvm_rs code first reserves memory by `Allocator`. (1MB maybe?)
- This reserved memory is not actually consumes RAM
- But when `node_to_bytes` is called, reserved memory is "materialized" and consumes RAM

I'm still looking for the actual problem and fix for now.

Reply:
no, the issue can be illustrated by this program:
```
(mod ()
(defun make_list (leaf n)
  (if n (c leaf (make_list leaf (- n 1))) leaf)
)

(make_list (lsh (q . 1) (q . 65535)) (q . 1000))
)
```

This creates one (somewhat) large atom, then builds a list of 1000 elements with that one atom. In memory, there's only a single atom, that's referenced by every cons-box/pair in the list. But if you serialize the list, you'll make a copy of that atom for every element in the list. i.e. you blow up memory usage.

The way we solved this in the interface to python is that we don't return a the resulting tree as actual python `SExp` objects, we return a native rust object that provides a view into the nodes allocated in the corresponding allocator. 

And in the critical parts we don't even return that, we parse the generator program output directly in rust and validates some of it immediately.

Reply:
I'm investigating whether `wasm-pack` can handle something like `LazyNode` in Python wheel.
As for the DoS issue, I think even after implementing `LazyNode` in wasm, the DoS attack can be effective everytime `serialization` is required.
So implementing `LazyNode` seems to just lower the chance of the attack and not completely preventing it.

Reply:
right, whoever serializes a structure has to trust it. I don't believe we ever do though, we serialize *programs* we don't necessarily trust, not their output.

Also, aren't all programs we run in the wallet trusted? Maybe the wasm bindings are meant for other applications than wallet.

Reply:
Oh I'm completely convinced. We serialize _program_, not its _output_.
The last remaining issue is to implement wasm version of `LazyNode`. I'll continue investigaing on that.

Reply:
@arvidn 
I've finally managed to implement wasm version of `LazyNode`.

This is a sample test code.
(You need to have NodeJS installed)

Prepare
```bash
git clone https://github.com/Chia-Network/clvm_rs
cd clvm_rs
wasm-pack build --release --target=nodejs wasm
node
```
(The code below is in the context of NodeJS)
```javascript
const wasm = require("./wasm/pkg/clvm_wasm.js");

// (q . 127)
prog = Uint8Array.from(Buffer.from("ff017f", "hex"));
arg = new Uint8Array([0]);

result = wasm.run_chia_program(prog, arg, BigInt("100000000000"), 0);
cost = result[0];
sexp = result[1];
console.log(cost, sexp, sexp.atom.toString() === "127");
```